cpp/arcticdb/version/version_core.hpp

```diff
@@ -209,5 +210,101 @@ folly::Future<ReadVersionOutput> read_frame_for_version(
 
+namespace arcticdb {
+
+struct Error {
+
+    explicit Error(folly::Function<void(std::string)> raiser, std::string msg);
+    void throw_error();
+
+    folly::Function<void(std::string)> raiser_;
+    std::string msg_;
+};
+
+using CheckOutcome = std::variant<Error, std::monostate>;
+using StaticSchemaCompactionChecks = folly::Function<CheckOutcome(const StreamDescriptor&, const pipelines::PipelineContext* const)>;
+using CompactionWrittenKeys = std::vector<VariantKey>;
+using CompactionResult = std::variant<CompactionWrittenKeys, Error>;
+
+void remove_written_keys(Store* store, CompactionWrittenKeys&& written_keys);
+
+template <typename IndexType, typename SchemaType, typename SegmentationPolicy, typename DensityPolicy, typename IteratorType>
+[[nodiscard]] CompactionResult do_compact(
+    IteratorType to_compact_start,
+    IteratorType to_compact_end,
+    const std::shared_ptr<pipelines::PipelineContext>& pipeline_context,
+    std::vector<pipelines::FrameSlice>& slices,
+    const std::shared_ptr<Store>& store,
+    bool convert_int_to_float,
+    std::optional<size_t> segment_size,
+    bool validate_index,
+    StaticSchemaCompactionChecks&& checks) {
+    CompactionResult result;
+    auto index = stream::index_type_from_descriptor(pipeline_context->descriptor());
+
+    std::vector<folly::Future<VariantKey>> write_futures;
+
+    stream::SegmentAggregator<IndexType, SchemaType, SegmentationPolicy, DensityPolicy>
+        aggregator{
+        [&slices](pipelines::FrameSlice &&slice) {
+            slices.emplace_back(std::move(slice));
+        },
+        SchemaType{pipeline_context->descriptor(), index},
+        [&write_futures, &store, &pipeline_context](SegmentInMemory &&segment) {
+            auto local_index_start = IndexType::start_value_for_segment(segment);
+            auto local_index_end = pipelines::end_index_generator(IndexType::end_value_for_segment(segment));
+            stream::StreamSink::PartialKey
+                pk{KeyType::TABLE_DATA, pipeline_context->version_id_, pipeline_context->stream_id_, local_index_start, local_index_end};
+
+            // TODO We should apply back pressure to the work we are generating here, to bound memory use
+            write_futures.emplace_back(store->write(pk, std::move(segment)));
+        },
+        segment_size.has_value() ? SegmentationPolicy{*segment_size} : SegmentationPolicy{}
+    };
+
+    for (auto it = to_compact_start; it != to_compact_end; ++it) {
+        auto sk = [&it]() {
+            if constexpr (std::is_same_v<IteratorType, pipelines::PipelineContext::iterator>)
+                return it->slice_and_key();
+            else
+                return *it;
+        }();
+        if (sk.slice().rows().diff() == 0) {
+            continue;
+        }
+
+        const SegmentInMemory& segment = sk.segment(store);
+
+        if(validate_index && segment.descriptor().sorted() != SortedValue::ASCENDING && segment.descriptor().sorted() != SortedValue::UNKNOWN) {
```

Comment:
I'd be tempted here to move the conditions (besides validate _index) into a short well-named function.